Add crossfilter.isElementFiltered() to test if data is filtered

drarmstr opened the pull request:
I often have a need to test if a data element is filtered in or not.  Based on various situations it is often useful to consider or ignore filters from an arbitrary set of dimensions.  (Just to provide an example of this, let's say one chart is a timeline that is rendering some dimension D.  When rendering that timeline it is often useful to ignore filters from dimension D as well as from a time dimension)  This can be done by tracking filtering in shadow structures, recomputing the filters, or other means.  But, it is redundant when Crossfilter already has the necessary state computed and stored.  Unfortunately, Crossfilter hides this state in internal private closure variables.  So, this push request adds a method to test if an element is filtered based on an arbitrary set of dimensions to ignore their filter.

This push request is meant to start a discussion.  Using `_zero` and `_offset` internally are only one approach to get the functionality working quickly.  The maintainers surely have some vision for how this type of functionality may be better integrated with where they plan on taking Crossfilter?  With the change to support >32 dimensions it also complicated it slightly.  Tests can be added based on discussion.

A related topic for discussion that we can take elsewhere is that I would also love similar functionality for dimension keys.  That is, to see if the data represented by a dimension's keys are currently filtered in or not while ignoring filters from a specified set of dimensions.

derekperkins commented:
@drarmstr Have you taken a look at github.com/crossfilter/universe? @tannerlinsley built that on top of crossfilter to provide a better api and management of those filter/dimension states.

drarmstr commented:
@derekperkins Not yet.  But, if it's built on top of Crossfilter then wouldn't it also be unable to access the private closure variables of Crossfilter?

tannerlinsley commented:
Universe does and will provide some pretty slick stuff along the lines of state management, but at the moment, that granularity is restricted with the deepest level being group/filter management.  To know if a record is currently filtered or not is, as @drarmstr suggests, currently very difficult and requires a lot of duplicated effort.  Not directly related, but reminiscent of this problem is #6.  Providing more metal-level control for things like this I think are definitely necessary if we plan to grow the functionality of crossfilter.  Indeed, the 32+ dimensions complicates this current implementation a bit.

Given the use case, which I don't think is that uncommon, and also the flexibility/ease of use it could open up, I think this is definitely worth more discussion.  @esjewett, could you think of any direct use for this in any of your current implementations?  I'm trying to think of how a method like this could enhance Universe as well...

tannerlinsley commented:
I think writing some failing tests for this use case would be a great start, too.  Never hurts.

esjewett commented:
I think this could make sense, though I don't completely understand the use-case described above. I agree about the need for failing tests. I'd say 2 other things:

1. Food for thought: I find it problematic to have the index that this function takes be the internal index of the Crossfilter data array. I think it should probably be defined as the index of the array returned by crossfilter.all. Effectively, this is the same thing right now.
2. Take a look at crossfilter_bitarray.prototype.zeroExcept and crossfilter_bitarray.prototype.onlyExcept in array.js. I think we should define something at this level and then the API should probably be a relatively thin wrapper around that. Does that make sense?

tannerlinsley commented:
Agreed
On Wed, Mar 16, 2016 at 5:00 PM Ethan Jewett <notifications@github.com>
wrote:

> I think this could make sense, though I don't completely understand the
> use-case described above. I agree about the need for failing tests. I'd say
> 2 other things:
>
>    1. Food for thought: I find it problematic to have the index that this
>    function takes be the internal index of the Crossfilter data array. I think
>    it should probably be defined as the index of the array returned by
>    crossfilter.all. Effectively, this is the same thing right now.
>    2. Take a look at crossfilter_bitarray.prototype.zeroExcept and
>    crossfilter_bitarray.prototype.onlyExcept in array.js. I think we should
>    define something at this level and then the API should probably be a
>    relatively thin wrapper around that. Does that make sense?
>
> —
> You are receiving this because you were mentioned.
>
>
> Reply to this email directly or view it on GitHub
> <https://github.com/crossfilter/crossfilter/pull/32#issuecomment-197570693>
>

drarmstr commented:
@tannerlinsley @esjewett Ok, thanks.  I added testing and refactored to move some functionality to the crossfilter_bitarray abstraction.  How does that look?

esjewett commented on crossfilter.js:
```diff
@@ -582,6 +582,15 @@ crossfilter_bitarray.prototype.zeroExcept = function(n, offset, zero) {
+crossfilter_bitarray.prototype.zeroExceptMask = function(n, mask) {
+  var i, len;
+  for (i = 0, len = this.subarrays; i < len; ++i)
+    if (this[i][n] & mask[i]) return false;
```
Can you add braces to the `if` statement to align with the coding style in the rest of the file?

esjewett commented on src/array.js:
```diff
@@ -144,6 +144,15 @@ crossfilter_bitarray.prototype.zeroExcept = function(n, offset, zero) {
+crossfilter_bitarray.prototype.zeroExceptMask = function(n, mask) {
+  var i, len;
+  for (i = 0, len = this.subarrays; i < len; ++i)
+    if (this[i][n] & mask[i]) return false;
```
Can you add braces to the if statement to align with the coding style in the rest of the file?

drarmstr replied:
Done

esjewett commented on src/crossfilter.js:
```diff
@@ -68,6 +69,22 @@ function crossfilter() {
+  // Return true if the data element at index i is filtered IN.
+  // Optionally, ignore the filters of any dimensions in the ignore_dimensions list.
+  function isElementFiltered(i, ignore_dimensions) {
+      var n,
```
I think we are generally using 2-space indentation. If you can switch to that for this change, it would be helpful. (In this file and in the test.)

drarmstr replied:
Done

esjewett commented on src/crossfilter.js:
```diff
@@ -121,6 +138,11 @@ function crossfilter() {
     zero = ~one;
+      
+    // Expose the dimension mask for internal use
+    dimension._offset = offset;
```
Is there a way we can avoid exposing this externally? I'd like to keep the external API clean.

drarmstr replied:
@esjewett Yeah, I thought this would prompt some discussion.  I'm not sure exactly how you would like this done to best fit with your particular vision.

offset, one, and zero were currently hidden as private closure variables in the dimension, so I couldn't access them from a higher-level crossfilter method.  They need to be exposed somehow from the dimension itself.  Adding an API or other approach would seem to expose an interface for this internal implementation detail, so that seemed wrong.  Thus the convention of using _ to denote a private member seemed like the easiest alternative.  I'll be happy to change this if you have a cleaner vision how to expose dimension-level state to crossfilter-level methods?

esjewett replied:
@drarmstr I don't have a clean approach here. I would say potentially create a map at the Crossfilter scope level with the dimension object as the key and store the offsets there as well as in the dimension scope. I believe they are just defined once when the dimension is created, so you would not have to worry about them changing. The in the isElementFiltered function you could use that map to get the necessary information.

drarmstr replied:
Well, since map keys are coerced to strings we would need a unique string key for each dimension then.  And since the offset/one/zero is defined in the dimension constructor itself it would also mean that map in the crossfilter scope would need to be exposed.  So, it seemed cleaner to expose private offset/zero/one in the dimension then expose the private offset/zero/one map in the crossfilter scope...

tannerlinsley replied:
I vote exposing it only on the dimension object with _ as the prefix
On Fri, Mar 18, 2016 at 12:47 PM Douglas Armstrong <notifications@github.com>
wrote:

> In src/crossfilter.js
> <https://github.com/crossfilter/crossfilter/pull/32#discussion_r56695490>:
>
> > @@ -121,6 +138,11 @@ function crossfilter() {
> >      offset = tmp.offset;
> >      one = tmp.one;
> >      zero = ~one;
> > +
> > +    // Expose the dimension mask for internal use
> > +    dimension._offset = offset;
>
> Well, since map keys are coerced to strings we would need a unique string
> key for each dimension then. And since the offset/one/zero is defined in
> the dimension constructor itself it would also mean that map in the
> crossfilter scope would need to be exposed. So, it seemed cleaner to expose
> private offset/zero/one in the dimension then expose the private
> offset/zero/one map in the crossfilter scope...
>
> —
> You are receiving this because you were mentioned.
>
>
> Reply to this email directly or view it on GitHub
> <https://github.com/crossfilter/crossfilter/pull/32/files/9884cc984fa733f5b9760f57317a5aba7369280e#r56695490>
>

esjewett replied:
@tannerlinsley That means it will be accessible on the external dimension API. I'd really prefer to keep the external API minimal. I think the original CF team did a really good job of that.

@drarmstr You're right - I had a brain malfunction :-) You'd need to maintain an array of dimension object references, and corresponding arrays of dimension offsets, etc. Then do an indexOf lookup on the dimension array and grab the corresponding index from the offset array. You'd want to maintain the arrays (splice) when dimension.dispose is called. If you want, I can do this after accepting the pull request onto a branch. Just let me know.

tannerlinsley replied:
Good point. Either way I'm good
On Fri, Mar 18, 2016 at 3:18 PM Ethan Jewett <notifications@github.com>
wrote:

> In src/crossfilter.js
> <https://github.com/crossfilter/crossfilter/pull/32#discussion_r56716138>:
>
> > @@ -121,6 +138,11 @@ function crossfilter() {
> >      offset = tmp.offset;
> >      one = tmp.one;
> >      zero = ~one;
> > +
> > +    // Expose the dimension mask for internal use
> > +    dimension._offset = offset;
>
> @tannerlinsley <https://github.com/tannerlinsley> That means it will be
> accessible on the external dimension API. I'd really prefer to keep the
> external API minimal. I think the original CF team did a really good job of
> that.
>
> @drarmstr <https://github.com/drarmstr> You're right - I had a brain
> malfunction :-) You'd need to maintain an array of dimension object
> references, and corresponding arrays of dimension offsets, etc. Then do an
> indexOf lookup on the dimension array and grab the corresponding index from
> the offset array. You'd want to maintain the arrays (splice) when
> dimension.dispose is called. If you want, I can do this after accepting the
> pull request onto a branch. Just let me know.
>
> —
> You are receiving this because you were mentioned.
>
>
> Reply to this email directly or view it on GitHub
> <https://github.com/crossfilter/crossfilter/pull/32/files/9884cc984fa733f5b9760f57317a5aba7369280e#r56716138>
>

gordonwoodhull replied:
This is tough. I'm with @esjewett that we should not expose internal variables just to make them accessible within crossfilter.

Another possible approach is to expose an identifier for the dimensions (`offset` is a unique integer I think?), and then keep an array of zeroes and ones. 

IOW add a method `dimension.id()` which we say is unique for all current dimensions. Secretly it is actually `offset`.

I don't understand this:
> And since the offset/one/zero is defined in the dimension constructor itself it would also mean that map in the crossfilter scope would need to be exposed.

drarmstr replied:
I would be against using indexOf to look up the entry, since that can be a performance concern and we wouldn't want that on a performance critical path.

(Ignore my other comment on exposing the map at the crossfilter scope, I didn't realize the dimension was in the crossfilter closure lexical scope.)

Therefore, exposing an id or key property for the dimension could work.  I could use that to look up in a map.  Or, actually, just programatically compute the mask based on the ID.  This would avoid the need for the map lookup.  And I could just set the id property in the dimension construction based on the provided offset/one values.

gordonwoodhull replied:
Don't think you need to set anything, I think you just need to add an `id()` method returning `offset` - effectively the closure is the dimension object here, so `offset` sticks around as long as the dimension does.

drarmstr replied:
@gordonwoodhull @tannerlinsley @esjewett Ok, I updated the push request to remove exposing dimension._offset/_zero/_one and replace it with a cleaner dimension ID.

esjewett commented:
Sweet - thanks for doing that. It looks good to me now with the exception of the specific comments I noted above.

esjewett commented:
So is the plan to merge this with the ID approach? I'd be ok with that. If this is being used in high-volume loops, that is the more efficient approach.

gordonwoodhull commented:
1. I think it should be an id method rather than a field, for consistency. We wouldn't want people to set it either.
 2. I wish it didn't need quite as much bit twiddling. I hadn't realized that the offset was just part of the number. 

I know the code has this stuff everywhere but it makes it a little hard to maintain. (The recent always-widening bug seems like a symptom of that.)

drarmstr commented:
@gordonwoodhull Ok, I converted `id` to an accessor method, though that does add another function call to the critical path.  And yes, there is some bit twiddling, but remember the point of Crossfilter is to be as fast as possible.  Most of the complexity here is due to the support for 32+ dimensions, which I have separate concerns for its performance regression.

esjewett commented:
I think a dictionary lookup of the offset might be faster than this bit-twiddling, but we'd have to test that. In any case, I agree that a method ID is the right way to go. It does introduce overhead, but it abstracts the internal value and that's important. With this level of abstraction, we can switch out the exact method of deriving the offset from the ID.

On the 32+ support, we can discuss that elsewhere, such as #31. I'd just say that performance isn't all that matters and I think that support for more dimensions itself is quite valuable. If there is a way we can speed it up, I'd love to see suggestions.

drarmstr commented:
So does this request look ok for merging now with the changes made?

gordonwoodhull commented:
LGTM. I have concerns about maintaining code with more and more bit twiddling as we add features, but we don't have to worry about that now. 

(Are function calls actually still a problem in compiled JavaScript? I would think they would get inlined, but I'm not an expert in JS efficiency.)

esjewett commented:
Sorry, I stepped on this with a pull request yesterday. I can merge it via command line today or you can update the pull request so it can be merged on Github.

Re bit twiddling: AFAICT, it's all to encode the offset and one mask into the id and decode it. Is that right? I would prefer to do this using a generated id/uuid and a map lookup. Am planning to do this after merging.

drarmstr commented:
@esjewett Have you had a chance to look at the conflicts?  Did you want me to try and resolve them?

esjewett commented:
Apologies for the delay. I will try to merge now...

drarmstr commented:
Thank you!  What is the procedure for updating the documentation?  Since this isn't in a stable release yet I gather it shouldn't be updated yet?

esjewett commented:
No problem, sorry again for the delay! If you could update the wiki regarding the API, that would be great. Just make a note that this functionality will be available in version 1.4.

drarmstr commented:
Ok, will do.
